<a href="https://colab.research.google.com/github/plee0617/IS-170-Binder/blob/Association-Rules/LeeIS170_MW_LAB5_Association_Rules.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# Pang Yia Lee

In [ ]:
import pandas as pd
import numpy as np
from mlxtend.frequent_patterns import apriori, association_rules

In [ ]:
df=pd.read_csv('GroceryStoreDataSet.csv')
df.head()

MILK,BREAD,BISCUIT
0  BREAD,MILK,BISCUIT,CORNFLAKES
1            BREAD,TEA,BOURNVITA
2           JAM,MAGGI,BREAD,MILK
3              MAGGI,TEA,BISCUIT
4            BREAD,TEA,BOURNVITA

In [ ]:
#Since it's missing a header we'll create one, the new column header 'Products' 
df=pd.read_csv('GroceryStoreDataSet.csv', names=['products'])
df.head() 


products
0             MILK,BREAD,BISCUIT
1  BREAD,MILK,BISCUIT,CORNFLAKES
2            BREAD,TEA,BOURNVITA
3           JAM,MAGGI,BREAD,MILK
4              MAGGI,TEA,BISCUIT

In [ ]:
df.shape

(20, 1)

In [ ]:
#following code converts a pandas dataframe into a list of lists.
data=list(df["products"].apply(lambda x:x.split(",")))
data

[['MILK', 'BREAD', 'BISCUIT'],
 ['BREAD', 'MILK', 'BISCUIT', 'CORNFLAKES'],
 ['BREAD', 'TEA', 'BOURNVITA'],
 ['JAM', 'MAGGI', 'BREAD', 'MILK'],
 ['MAGGI', 'TEA', 'BISCUIT'],
 ['BREAD', 'TEA', 'BOURNVITA'],
 ['MAGGI', 'TEA', 'CORNFLAKES'],
 ['MAGGI', 'BREAD', 'TEA', 'BISCUIT'],
 ['JAM', 'MAGGI', 'BREAD', 'TEA'],
 ['BREAD', 'MILK'],
 ['COFFEE', 'COCK', 'BISCUIT', 'CORNFLAKES'],
 ['COFFEE', 'COCK', 'BISCUIT', 'CORNFLAKES'],
 ['COFFEE', 'SUGER', 'BOURNVITA'],
 ['BREAD', 'COFFEE', 'COCK'],
 ['BREAD', 'SUGER', 'BISCUIT'],
 ['COFFEE', 'SUGER', 'CORNFLAKES'],
 ['BREAD', 'SUGER', 'BOURNVITA'],
 ['BREAD', 'COFFEE', 'SUGER'],
 ['BREAD', 'COFFEE', 'SUGER'],
 ['TEA', 'MILK', 'COFFEE', 'CORNFLAKES']]

In [ ]:
from mlxtend.preprocessing import TransactionEncoder
a=TransactionEncoder()
a_data=a.fit(data).transform(data)
df=pd.DataFrame(a_data, columns=a.columns_)

#Turn False & True to 0 & 1 for easy readability
df=df.replace(False,0)
df=df.replace(True,1)

df
                

BISCUIT  BOURNVITA  BREAD  COCK  COFFEE  CORNFLAKES  JAM  MAGGI  MILK  \
0         1          0      1     0       0           0    0      0     1   
1         1          0      1     0       0           1    0      0     1   
2         0          1      1     0       0           0    0      0     0   
3         0          0      1     0       0           0    1      1     1   
4         1          0      0     0       0           0    0      1     0   
5         0          1      1     0       0           0    0      0     0   
6         0          0      0     0       0           1    0      1     0   
7         1          0      1     0       0           0    0      1     0   
8         0          0      1     0       0           0    1      1     0   
9         0          0      1     0       0           0    0      0     1   
10        1          0      0     1       1           1    0      0     0   
11        1          0      0     1       1           1    0      0     0   
12        0          1      0     0       1           0    0      0     0   
13        0          0      1     1       1           0    0      0     0   
14        1          0      1     0       0           0    0      0     0   
15        0          0      0     0       1           1    0      0     0   
16        0          1      1     0       0           0    0      0     0   
17        0          0      1     0       1           0    0      0     0   
18        0          0      1     0       1           0    0      0     0   
19        0          0      0     0       1           1    0      0     1   

    SUGER  TEA  
0       0    0  
1       0    0  
2       0    1  
3       0    0  
4       0    1  
5       0    1  
6       0    1  
7       0    1  
8       0    1  
9       0    0  
10      0    0  
11      0    0  
12      1    0  
13      0    0  
14      1    0  
15      1    0  
16      1    0  
17      1    0  
18      1    0  
19      0    1

In [ ]:
df=apriori(df,min_support=0.2,use_colnames=True)
df

support              itemsets
0      0.35             (BISCUIT)
1      0.20           (BOURNVITA)
2      0.65               (BREAD)
3      0.40              (COFFEE)
4      0.30          (CORNFLAKES)
5      0.25               (MAGGI)
6      0.25                (MILK)
7      0.30               (SUGER)
8      0.35                 (TEA)
9      0.20      (BISCUIT, BREAD)
10     0.20         (BREAD, MILK)
11     0.20        (SUGER, BREAD)
12     0.20          (TEA, BREAD)
13     0.20  (COFFEE, CORNFLAKES)
14     0.20       (COFFEE, SUGER)
15     0.20          (MAGGI, TEA)

In [ ]:
df_ar=association_rules(df,metric="lift", min_threshold=1)
df_ar

antecedents   consequents  antecedent support  consequent support  \
0       (BREAD)        (MILK)                0.65                0.25   
1        (MILK)       (BREAD)                0.25                0.65   
2       (SUGER)       (BREAD)                0.30                0.65   
3       (BREAD)       (SUGER)                0.65                0.30   
4      (COFFEE)  (CORNFLAKES)                0.40                0.30   
5  (CORNFLAKES)      (COFFEE)                0.30                0.40   
6      (COFFEE)       (SUGER)                0.40                0.30   
7       (SUGER)      (COFFEE)                0.30                0.40   
8       (MAGGI)         (TEA)                0.25                0.35   
9         (TEA)       (MAGGI)                0.35                0.25   

   support  confidence      lift  leverage  conviction  
0      0.2    0.307692  1.230769    0.0375    1.083333  
1      0.2    0.800000  1.230769    0.0375    1.750000  
2      0.2    0.666667  1.025641    0.0050    1.050000  
3      0.2    0.307692  1.025641    0.0050    1.011111  
4      0.2    0.500000  1.666667    0.0800    1.400000  
5      0.2    0.666667  1.666667    0.0800    1.800000  
6      0.2    0.500000  1.666667    0.0800    1.400000  
7      0.2    0.666667  1.666667    0.0800    1.800000  
8      0.2    0.800000  2.285714    0.1125    3.250000  
9      0.2    0.571429  2.285714    0.1125    1.750000

In [ ]:
df_ar=association_rules(df,metric="confidence", min_threshold=0.6)
df_ar


antecedents consequents  antecedent support  consequent support  support  \
0        (MILK)     (BREAD)                0.25                0.65      0.2   
1       (SUGER)     (BREAD)                0.30                0.65      0.2   
2  (CORNFLAKES)    (COFFEE)                0.30                0.40      0.2   
3       (SUGER)    (COFFEE)                0.30                0.40      0.2   
4       (MAGGI)       (TEA)                0.25                0.35      0.2   

   confidence      lift  leverage  conviction  
0    0.800000  1.230769    0.0375        1.75  
1    0.666667  1.025641    0.0050        1.05  
2    0.666667  1.666667    0.0800        1.80  
3    0.666667  1.666667    0.0800        1.80  
4    0.800000  2.285714    0.1125        3.25

From the result above, the two items which is milk and bread has a 0.25 and 0.65 and a confidence of .80, which means there is an 80% chance that if a person bought milk, they will also buy bread.